File: packages/belfast/src/core/Buffer.ts
import type { Device } from "./Device";
import { writeBuffer } from "./GPUResources";

export const BufferUsage = {
  vertex: GPUBufferUsage.VERTEX | GPUBufferUsage.COPY_DST,
  storage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_DST,
  vertexStorage: GPUBufferUsage.VERTEX | GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_DST,
} as const;

export class Buffer {
  readonly gpu: GPUBuffer;
  readonly size: number;
  readonly usage: GPUBufferUsageFlags;
  readonly label?: string;

  private constructor(gpu: GPUBuffer, size: number, usage: GPUBufferUsageFlags, label?: string) {
    this.gpu = gpu;
    this.size = size;
    this.usage = usage;
    this.label = label;
  }

  static create(device: Device, size: number, usage: GPUBufferUsageFlags, label?: string): Buffer {
    const gpu = device.device.createBuffer({ size, usage, label });
    return new Buffer(gpu, size, usage, label);
  }

  static fromData(
    device: Device,
    data: ArrayBuffer | ArrayBufferView,
    usage: GPUBufferUsageFlags,
    label?: string,
  ): Buffer {
    const byteLength = data.byteLength;
    const buffer = Buffer.create(device, byteLength, usage, label);
    buffer.write(device, data);
    return buffer;
  }

  write(device: Device, data: ArrayBuffer | ArrayBufferView, byteOffset = 0): void {
    writeBuffer(device, this.gpu, data, byteOffset);
  }

  destroy(): void {
    this.gpu.destroy();
  }
}

File: crates/belfast/src/buffer.rs
use std::sync::Arc;

use wgpu::util::DeviceExt;

use crate::Device;

#[derive(Clone)]
pub struct Buffer {
    gpu: Arc<wgpu::Buffer>,
    size: u64,
}

impl Buffer {
    pub fn create(device: &Device, size: u64, usage: wgpu::BufferUsages, label: &str) -> Self {
        let gpu = device.gpu().create_buffer(&wgpu::BufferDescriptor {
            label: Some(label),
            size,
            usage,
            mapped_at_creation: false,
        });
        Self {
            gpu: Arc::new(gpu),
            size,
        }
    }

    pub fn from_data<T: bytemuck::Pod>(
        device: &Device,
        data: &[T],
        usage: wgpu::BufferUsages,
        label: &str,
    ) -> Self {
        let contents = bytemuck::cast_slice(data);
        let gpu = device
            .gpu()
            .create_buffer_init(&wgpu::util::BufferInitDescriptor {
                label: Some(label),
                contents,
                usage,
            });
        Self {
            gpu: Arc::new(gpu),
            size: contents.len() as u64,
        }
    }

    pub fn write<T: bytemuck::Pod>(&self, device: &Device, data: &[T], byte_offset: u64) {
        device
            .queue()
            .write_buffer(&self.gpu, byte_offset, bytemuck::cast_slice(data));
    }

    pub fn gpu(&self) -> &wgpu::Buffer {
        &self.gpu
    }

    pub fn size(&self) -> u64 {
        self.size
    }
}

pub struct BufferUsage;

impl BufferUsage {
    pub fn vertex() -> wgpu::BufferUsages {
        wgpu::BufferUsages::VERTEX | wgpu::BufferUsages::COPY_DST
    }

    pub fn index() -> wgpu::BufferUsages {
        wgpu::BufferUsages::INDEX | wgpu::BufferUsages::COPY_DST
    }

    pub fn uniform() -> wgpu::BufferUsages {
        wgpu::BufferUsages::UNIFORM | wgpu::BufferUsages::COPY_DST
    }

    pub fn storage() -> wgpu::BufferUsages {
        wgpu::BufferUsages::STORAGE | wgpu::BufferUsages::COPY_DST
    }

    pub fn vertex_storage() -> wgpu::BufferUsages {
        wgpu::BufferUsages::VERTEX | wgpu::BufferUsages::STORAGE | wgpu::BufferUsages::COPY_DST
    }
}
